ctapipe/tools/make_irf.py

```diff
@@ -0,0 +1,385 @@
+            ("gamma", self.gamma_file, PYIRF_SPECTRA[self.gamma_sim_spectrum]),
+            ("proton", self.proton_file, PYIRF_SPECTRA[self.proton_sim_spectrum]),
+            (
+                "electron",
+                self.electron_file,
+                PYIRF_SPECTRA[self.electron_sim_spectrum],
+            ),
+        ]:
+            with TableLoader(file, **opts) as load:
+                Provenance().add_input_file(file)
+                header = self.epp.make_empty_table()
+                sim_info, spectrum, obs_conf = self.get_metadata(load)
+                if kind == "gamma":
+                    self.sim_info = sim_info
+                    self.spectrum = spectrum
+                bits = [header]
+                for start, stop, events in load.read_subarray_events_chunked(
+                    self.chunk_size
+                ):
+                    selected = self.epp.normalise_column_names(events)
+                    selected = selected[self.epp.get_table_mask(selected)]
+                    selected = self.make_derived_columns(
+                        kind, selected, spectrum, target_spectrum, obs_conf
+                    )
+                    bits.append(selected)
+
+                table = vstack(bits, join_type="exact")
+                reduced_events[kind] = table
+
+        select_ON = reduced_events["gamma"]["theta"] <= self.ON_radius * u.deg
+        self.signal = reduced_events["gamma"][select_ON]
+        self.background = vstack([reduced_events["proton"], reduced_events["electron"]])
+
+    def setup(self):
+        self.co = CutOptimising(parent=self)
+        self.e_bins = OutputEnergyBinning(parent=self)
+        self.bins = DataBinning(parent=self)
+        self.epp = EventPreProcessor(parent=self)
+
+        self.reco_energy_bins = self.e_bins.reco_energy_bins()
+        self.true_energy_bins = self.e_bins.true_energy_bins()
+        self.energy_migration_bins = self.e_bins.energy_migration_bins()
+
+        self.source_offset_bins = self.bins.source_offset_bins()
+        self.fov_offset_bins = self.bins.fov_offset_bins()
+        self.bkg_fov_offset_bins = self.bins.bkg_fov_offset_bins()
+
+    def start(self):
+        self.load_preselected_events()
+        self.gh_cuts, self.theta_cuts_opt, sens2 = self.co.optimise_gh_cut(
+            self.signal,
+            self.background,
+            self.alpha,
+            self.max_bg_radius,
+        )
+
+        self.signal["selected_theta"] = evaluate_binned_cut(
+            self.signal["theta"],
+            self.signal["reco_energy"],
+            self.theta_cuts_opt,
+            operator.le,
+        )
+        self.signal["selected"] = (
+            self.signal["selected_theta"] & self.signal["selected_gh"]
+        )
+
+        # calculate sensitivity
+        signal_hist = create_histogram_table(
+            self.signal[self.signal["selected"]], bins=self.reco_energy_bins
+        )
+
+        background_hist = estimate_background(
```

Comment:
What is the difference between this background hist and the one generated in `finish()`? One uses `estimate_background` and one uses `background_2d`.  Are these the same computations? I worry that things are computed multiple times, and it's not clear which ones are in the output vs used in the sensitivity

Reply:
In fact, I don't find `pyirf.sensitivity.estimate_background` in the PyIRF documentation, so could it be incorrect?

Reply:
That function is what pyIRF uses [internally](https://github.com/cta-observatory/pyirf/blob/12a609566230d01a68a822aad38f080b60c473ce/pyirf/cut_optimization.py#L119) so probably this is a problem with the docs. 

To be honest I'm not sure what the difference is. It looks like [estimate_background](https://github.com/cta-observatory/pyirf/blob/main/pyirf/sensitivity.py#L268) only does one offset bin but also rescales from the size of the Off region to the On region. Meanwhile [background_2d](https://github.com/cta-observatory/pyirf/blob/main/pyirf/irf/background.py) calculates background rates in each offset bin, but does not rescale to the On region size...

Reply:
`pyirf.sensitivity.estimate_background`

this function estimates the number of background *counts*  (n_off) for the given observation time and given on region size in the binning used for the IRF computation. 

 
`pyirf.irf.background.background_2d` is the background IRF component as defined by GADF.